### PR TITLE
Archive unpacking

### PR DESCRIPTION
**Pull request by cnr**
This adds `--unpack-archives` as a command switch to `fossa analyze`. Discovered `.tar`, `.tar.gz`, `.zip`, `.jar`, and `.rpm` files will be recursively unarchived and analyzed.

This required using absolute filepaths in all strategies, so there are a bunch of "unrelated" changes in this PR. This is required because archives are unpacked in a temporary directory, and the current working directory is process-global -- so reading relative files/directories won't work if we want to run things in parallel.

### CODE REVIEW

File: src/Control/Effect/Finally.hs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module Control.Effect.Finally
```
Comment (cnr) [inline src/Control/Effect/Finally.hs:1]: We add a new effect here -- `Finally` -- that has one operation:
```hs
onExit :: Has Finally sig m => m a -> m ()
```

This is used to clean up unpacked archive directories on program exit. We'd normally eagerly clean up these directories after discovering and running strategies, but we don't have a way to tell when the recursively-discovered strategies complete.

File: src/Strategy/Archive/RPM.hs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import qualified Data.CPIO as CPIO
+import Path
+import Prelude
+import qualified Path.IO as PIO
+
+extractRpm :: Has (Lift IO) sig m => Path Abs Dir -> Path Abs File -> m ()
+extractRpm dir rpmFile = do
+  res <-
+    sendIO . runResourceT . runExceptT . runConduit $
+      readRPMEntries rpmFile .| filterC (not . CPIO.isEntryDirectory) .| sinkDir
```
Comment (cnr) [inline src/Strategy/Archive/RPM.hs:23]: This uses a streaming library called `conduit` -- you can think of the `.|` operators as "unix pipes".

As for how to read the conduit type signatures: `ConduitT i o m a` has:

- "streaming input type" `i`
- "streaming output type" `o`
- underlying monad `m` (it's a monad transformer)
- result type `a` (like any other monad)

File: src/Control/Effect/Finally.hs (hunk 1 above)
Comment (scruffystuffs) [reply src/Control/Effect/Finally.hs:1]: Bummer.  Staggering those potentially large IO ops could be a lot more performant.  What's the exact reason we don't know when they're done?

Comment (cnr) [reply src/Control/Effect/Finally.hs:1]: The discover actions don't keep track of the strategies they find -- they just fork the strategy as a new task to be run in parallel and forget about it. It's partially a consequence of how aggressively parallel we are

File: src/App/Fossa/Main.hs [hunk 3]
```diff
@@ -117,6 +117,7 @@ analyzeOpts :: Parser AnalyzeOptions
 analyzeOpts =
   AnalyzeOptions
     <$> switch (long "output" <> short 'o' <> help "Output results to stdout instead of uploading to fossa")
+    <*> switch (long "unpack-archives" <> help "Recursively unpack and analyze discovered archives")
```
Comment (zlav) [inline src/App/Fossa/Main.hs:120]: I like the name `unpack-archives`, I think it is good to be explicit here.

File: src/Strategy/Archive.hs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  (Has (Lift IO) sig m, Has Finally sig m) =>
+  -- | Archive extraction function
+  (Path Abs Dir -> Path Abs File -> m ()) ->
+  -- | Path to archive
+  Path Abs File ->
+  -- | Callback
+  (Path Abs Dir -> m ()) ->
+  m ()
+withArchive extract file go = do
+  tmpDir <- mkTempDir (fileName file)
```
Comment (zlav) [inline src/Strategy/Archive.hs:67]: Does this not cause a collision between the filename and the newly created temp directory? I would think that we have to strip the file extension before creating the directory?

Comment (cnr) [reply src/Strategy/Archive.hs:67]: The temp directory is created as a _system_-level temp directory (like in `/tmp`), and the name provided serves as a "template" or prefix that'll be used when creating the system temp directory.

If a folder with that name already exists, it'll arbitrarily append random garbage to the end of it

**Pull request merged**